File: src/omnia/data/structure/observable/writable/WritableObservableMapImpl.java
package omnia.data.structure.observable.writable;

import static java.util.Objects.requireNonNull;
import static omnia.data.stream.Collectors.toSet;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import java.util.Objects;
import java.util.Optional;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;
import omnia.data.structure.Collection;
import omnia.data.structure.Map;
import omnia.data.structure.Set;
import omnia.data.structure.immutable.ImmutableMap;
import omnia.data.structure.immutable.ImmutableSet;
import omnia.data.structure.observable.ObservableMap;

final class WritableObservableMapImpl<K, V> implements WritableObservableMap<K, V> {

  private volatile ImmutableMap<K, V> currentState;
  private final FlowableProcessor<MutationEvent> mutationEventProcessor =
      PublishProcessor.<MutationEvent>create().toSerialized();

  WritableObservableMapImpl() {
    currentState = ImmutableMap.empty();
  }

  WritableObservableMapImpl(Map<? extends K, ? extends V> other) {
    currentState = ImmutableMap.copyOf(other);
  }

  @Override
  public void putMapping(K key, V value) {
    mutateState(
        currentState -> currentState.valueOf(key).map(v -> !Objects.equals(v, value)).orElse(true),
        currentState -> currentState.toBuilder().putMapping(key, value).build(),
        (previousState, newState) ->
            ImmutableSet.of(
                previousState.valueOf(key)
                    .<MapOperation<K, V>>map(previous -> new ReplaceInMap<>(key, previous, value))
                    .orElseGet(() -> new AddToMap<K, V>(key, value))));
  }

  @Override
  public V putMappingIfAbsent(K key, Supplier<V> value) {
    synchronized (this) {
      mutateState(
          currentState -> currentState.valueOf(key).isEmpty(),
          currentState -> currentState.toBuilder().putMapping(key, value.get()).build(),
          (previousState, newState) ->
              ImmutableSet.of(new AddToMap<>(key, newState.valueOf(key).get())));
      return currentState.valueOf(key).get();
    }
  }

  @Override
  public Optional<V> removeKey(K key) {
    synchronized (this) {
      Optional<V> removedValue = currentState.valueOf(key);
      mutateState(
          currentState -> currentState.valueOf(key).isPresent(),
          currentState -> currentState.toBuilder().removeKey(key).build(),
          (previousState, newState) ->
              ImmutableSet.of(new RemoveFromMap<>(key, previousState.valueOf(key).get())));
      return removedValue;
    }
  }

  @Override
  public Optional<V> removeUnknownTypedKey(Object key) {
    synchronized (this) {
      Optional<V> removedValue = currentState.valueOfUnknownTyped(key);
      mutateState(
          currentState -> currentState.valueOfUnknownTyped(key).isPresent(),
          currentState -> currentState.toBuilder().removeUnknownTypedKey(key).build(),
          (previousState, newState) ->
              ImmutableSet.<MapOperation<K, V>>of(
                  new RemoveFromMap<>((K) key, previousState.valueOfUnknownTyped(key).get())));
      return removedValue;
    }
  }

  private boolean mutateState(
      Predicate<? super ImmutableMap<K, V>> shouldChange,
      Function<? super ImmutableMap<K, V>, ? extends ImmutableMap<K, V>> mutateState,
      BiFunction<? super ImmutableMap<K, V>, ? super ImmutableMap<K, V>, ? extends Set<MapOperation<K, V>>> mutationOperations) {
    synchronized (this) {
      ImmutableMap<K, V> previousState = currentState;
      if (!shouldChange.test(previousState)) {
        return false;
      }
      ImmutableMap<K, V> nextState = requireNonNull(mutateState.apply(previousState));
      currentState = nextState;
      Set<MapOperation<K, V>> operations = mutationOperations.apply(previousState, nextState);
      mutationEventProcessor.onNext(new MutationEvent(nextState, operations));
      return true;
    }
  }

  @Override
  public Set<Entry<K, V>> entries() {
    return getState().entries();
  }

  @Override
  public Set<K> keys() {
    return getState().keys();
  }

  @Override
  public Collection<V> values() {
    return getState().values();
  }

  @Override
  public Optional<V> valueOf(K key) {
    return getState().valueOf(key);
  }

  @Override
  public Optional<V> valueOfUnknownTyped(Object key) {
    return getState().valueOfUnknownTyped(key);
  }

  @Override
  public Set<K> keysOf(V value) {
    return getState().keysOf(value);
  }

  @Override
  public Set<K> keysOfUnknownTyped(Object value) {
    return getState().keysOfUnknownTyped(value);
  }

  private ImmutableMap<K, V> getState() {
    synchronized (this) {
      return currentState;
    }
  }

  @Override
  public ObservableChannels observe() {
    return new ObservableChannels();
  }

  private final class ObservableChannels extends GenericObservableChannels<Map<K, V>, MutationEvent>
      implements ObservableMap.ObservableChannels<K, V> {
    private ObservableChannels() {
      super(
          Flowable.<Map<K, V>>create(
                  flowableEmitter -> {
                    flowableEmitter.onNext(getState());
                    flowableEmitter.onComplete();
                  },
                  BackpressureStrategy.BUFFER)
              .concatWith(mutationEventProcessor.map(MutationEvent::state)),
          Flowable.<MutationEvent>create(
                  flowableEmitter -> {
                    flowableEmitter.onNext(generateMutationEventForNewSubscription());
                    flowableEmitter.onComplete();
                  },
                  BackpressureStrategy.BUFFER)
              .concatWith(mutationEventProcessor));
    }
  }

  private MutationEvent generateMutationEventForNewSubscription() {
    Map<K, V> state = getState();
    return new MutationEvent(
        state,
        state.entries()
            .stream()
            .<MapOperation<K, V>>map(entry -> new AddToMap<>(entry.key(), entry.value()))
            .collect(toSet()));
  }

  private final class MutationEvent extends GenericMutationEvent<Map<K, V>, Set<MapOperation<K, V>>>
      implements ObservableMap.MutationEvent<K, V> {
    private MutationEvent(Map<K, V> state, Set<MapOperation<K, V>> operations) {
      super(state, operations);
    }
  }

  @Override
  public ObservableMap<K, V> toReadOnly() {
    return new ObservableMap<>() {
      @Override
      public ObservableChannels<K, V> observe() {
        return WritableObservableMapImpl.this.observe();
      }

      @Override
      public Set<K> keys() {
        return WritableObservableMapImpl.this.keys();
      }

      @Override
      public Collection<V> values() {
        return WritableObservableMapImpl.this.values();
      }

      @Override
      public Set<Entry<K, V>> entries() {
        return WritableObservableMapImpl.this.entries();
      }

      @Override
      public Optional<V> valueOfUnknownTyped(Object key) {
        return WritableObservableMapImpl.this.valueOfUnknownTyped(key);
      }

      @Override
      public Set<K> keysOfUnknownTyped(Object value) {
        return WritableObservableMapImpl.this.keysOfUnknownTyped(value);
      }
    };
  }

  private static final class AddToMap<K, V> implements ObservableMap.AddToMap<K, V> {
    private final K key;
    private final V value;

    private AddToMap(K key, V value) {
      this.key = key;
      this.value = value;
    }

    @Override
    public K key() {
      return key;
    }

    @Override
    public V value() {
      return value;
    }
  }

  private static final class RemoveFromMap<K, V> implements ObservableMap.RemoveFromMap<K, V> {
    private final K key;
    private final V value;

    private RemoveFromMap(K key, V value) {
      this.key = key;
      this.value = value;
    }

    @Override
    public K key() {
      return key;
    }

    @Override
    public V value() {
      return value;
    }
  }

  private static final class ReplaceInMap<K, V> implements ObservableMap.ReplaceInMap<K, V> {
    private final K key;
    private final V replacedValue;
    private final V newValue;

    private ReplaceInMap(K key, V replacedValue, V newValue) {
      this.key = key;
      this.replacedValue = replacedValue;
      this.newValue = newValue;
    }

    @Override
    public K key() {
      return key;
    }

    @Override
    public V replacedValue() {
      return replacedValue;
    }

    @Override
    public V newValue() {
      return newValue;
    }
  }
}

File: src/omnia/data/structure/observable/writable/WritableObservableMapImpl.kt
package omnia.data.structure.observable.writable

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableEmitter
import io.reactivex.rxjava3.subjects.PublishSubject
import io.reactivex.rxjava3.subjects.Subject
import java.util.Objects
import java.util.function.BiFunction
import java.util.function.Predicate
import java.util.function.Supplier
import omnia.data.stream.Collectors
import omnia.data.structure.Collection
import omnia.data.structure.Map
import omnia.data.structure.Set
import omnia.data.structure.immutable.ImmutableMap
import omnia.data.structure.immutable.ImmutableSet
import omnia.data.structure.observable.ObservableMap
import omnia.data.structure.observable.ObservableMap.MapOperation

internal class WritableObservableMapImpl<K : Any, V : Any> : WritableObservableMap<K, V> {

  @Volatile
  private var currentState: ImmutableMap<K, V>
  private val mutationEvents: Subject<MutationEvent> =
    PublishSubject.create<MutationEvent>().toSerialized()

  constructor() {
    currentState = ImmutableMap.empty()
  }

  constructor(other: Map<out K, out V>) {
    currentState = ImmutableMap.copyOf(other)
  }

  override fun putMapping(key: K, value: V) {
    mutateState(
      { currentState -> currentState.valueOf(key)?.let { it != value } ?: true },
      { currentState -> currentState.toBuilder().putMapping(key, value).build() }
    ) { previousState, _ ->
      ImmutableSet.of(
        previousState.valueOf(key)
            ?.let { ReplaceInMap(key, it, value) }
            ?: AddToMap(key, value))
    }
  }

  override fun putMappingIfAbsent(key: K, value: Supplier<V>): V {
    synchronized(this) {
      mutateState(
        { currentState -> currentState.valueOf(key) == null },
        { currentState -> currentState.toBuilder().putMapping(key, value.get()).build() }
      ) { _, newState -> ImmutableSet.of(AddToMap(key, newState.valueOf(key)!!)) }
      return currentState.valueOf(key)!!
    }
  }

  override fun removeKey(key: K): V? {
    synchronized(this) {
      val removedValue = currentState.valueOf(key)
      mutateState(
        { currentState -> currentState.valueOf(key) != null },
        { currentState -> currentState.toBuilder().removeKey(key).build() }
      ) { previousState, _ ->
        ImmutableSet.of(
          RemoveFromMap(
            key,
            previousState.valueOf(key)!!
          )
        )
      }
      return removedValue
    }
  }

  override fun removeUnknownTypedKey(key: Any?): V? {
    synchronized(this) {
      val removedValue = currentState.valueOfUnknownTyped(key)
      mutateState(
        { currentState -> currentState.valueOfUnknownTyped(key) != null },
        { currentState -> currentState.toBuilder().removeUnknownTypedKey(key).build() }
      ) { previousState, _ ->
        @Suppress("UNCHECKED_CAST")
        ImmutableSet.of(
          RemoveFromMap(
            key as K, previousState.valueOfUnknownTyped(key)!!
          )
        )
      }
      return removedValue
    }
  }

  private fun mutateState(
    shouldChange: Predicate<in ImmutableMap<K, V>>,
    mutateState: java.util.function.Function<in ImmutableMap<K, V>, out ImmutableMap<K, V>>,
    mutationOperations: BiFunction<in ImmutableMap<K, V>, in ImmutableMap<K, V>, out Set<MapOperation<K, V>>>
  ): Boolean {
    synchronized(this) {
      val previousState = currentState
      if (!shouldChange.test(previousState)) {
        return false
      }
      val nextState = Objects.requireNonNull(mutateState.apply(previousState))
      currentState = nextState
      val operations = mutationOperations.apply(previousState, nextState)
      mutationEvents.onNext(MutationEvent(nextState, operations))
      return true
    }
  }

  override fun entries(): Set<Map.Entry<K, V>> {
    return state.entries()
  }

  override fun keys(): Set<K> {
    return state.keys()
  }

  override fun values(): Collection<V> {
    return state.values()
  }

  override fun valueOf(key: K): V? {
    return state.valueOf(key)
  }

  override fun valueOfUnknownTyped(key: Any?): V? {
    return state.valueOfUnknownTyped(key)
  }

  override fun keysOf(value: V): Set<K> {
    return state.keysOf(value)
  }

  override fun keysOfUnknownTyped(value: Any?): Set<K> {
    return state.keysOfUnknownTyped(value)
  }

  private val state: ImmutableMap<K, V>
    get() {
      synchronized(this) { return currentState }
    }

  override fun observe(): ObservableChannels {
    return ObservableChannels()
  }

  inner class ObservableChannels : GenericObservableChannels<Map<K, V>, MutationEvent>(
    Observable.create { flowableEmitter: ObservableEmitter<Map<K, V>> ->
      flowableEmitter.onNext(state)
      flowableEmitter.onComplete()
    }
      .concatWith(mutationEvents.map { obj: MutationEvent -> obj.state() }),
    Observable.create { flowableEmitter: ObservableEmitter<MutationEvent> ->
      flowableEmitter.onNext(generateMutationEventForNewSubscription())
      flowableEmitter.onComplete()
    }
      .concatWith(mutationEvents)), ObservableMap.ObservableChannels<K, V>

  private fun generateMutationEventForNewSubscription(): MutationEvent {
    val state: Map<K, V> = state
    return MutationEvent(
      state,
      state.entries()
        .stream()
        .map<MapOperation<K, V>> { entry: Map.Entry<K, V> -> AddToMap(entry.key(), entry.value()) }
        .collect(Collectors.toSet()))
  }

  inner class MutationEvent(state: Map<K, V>, operations: Set<MapOperation<K, V>>) :
    GenericMutationEvent<Map<K, V>, Set<MapOperation<K, V>>>(state, operations),
    ObservableMap.MutationEvent<K, V>

  override fun toReadOnly(): ObservableMap<K, V> {
    return object : ObservableMap<K, V> {
      override fun observe(): ObservableMap.ObservableChannels<K, V> {
        return this@WritableObservableMapImpl.observe()
      }

      override fun keys(): Set<K> {
        return this@WritableObservableMapImpl.keys()
      }

      override fun values(): Collection<V> {
        return this@WritableObservableMapImpl.values()
      }

      override fun entries(): Set<Map.Entry<K, V>> {
        return this@WritableObservableMapImpl.entries()
      }

      override fun valueOfUnknownTyped(key: Any?): V? {
        return this@WritableObservableMapImpl.valueOfUnknownTyped(key)
      }

      override fun keysOfUnknownTyped(value: Any?): Set<K> {
        return this@WritableObservableMapImpl.keysOfUnknownTyped(value)
      }
    }
  }

  private class AddToMap<K : Any, V : Any>(private val key: K, private val value: V) :
    ObservableMap.AddToMap<K, V> {

    override fun key(): K {
      return key
    }

    override fun value(): V {
      return value
    }
  }

  private class RemoveFromMap<K : Any, V : Any>(private val key: K, private val value: V) :
    ObservableMap.RemoveFromMap<K, V> {

    override fun key(): K {
      return key
    }

    override fun value(): V {
      return value
    }
  }

  private class ReplaceInMap<K : Any, V : Any>(
    private val key: K,
    private val replacedValue: V,
    private val newValue: V
  ) : ObservableMap.ReplaceInMap<K, V> {

    override fun key(): K {
      return key
    }

    override fun replacedValue(): V {
      return replacedValue
    }

    override fun newValue(): V {
      return newValue
    }
  }
}